File: app/src/main/java/com/example/mytourguideapp/MainActivity.java
package com.example.mytourguideapp;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;
import androidx.viewbinding.ViewBindings;
import androidx.viewpager2.widget.ViewPager2;

import com.example.mytourguideapp.databinding.ActivityMainBinding;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

public class MainActivity extends AppCompatActivity {

    CategoryAdapter categoryAdapter;
    TabLayout tabLayout;
    ViewPager2 viewPager2;

    private ActivityMainBinding binding;

    private String[] titles = new String[]{"Mountains", "Cities", "Islands", "Seas"};


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        viewPager2 = findViewById(R.id.view_pager);
        tabLayout = findViewById(R.id.tab_layout);
        categoryAdapter = new CategoryAdapter(this);

        //set adapter viewpager
        viewPager2.setAdapter(categoryAdapter);
        new TabLayoutMediator(tabLayout, viewPager2,((tab, position) -> tab.setText(titles[position]))).attach();

    }

}

File: app/src/main/java/com/example/mytourguideapp/MainActivity.kt
package com.example.mytourguideapp

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager2.widget.ViewPager2
import com.example.mytourguideapp.databinding.ActivityMainBinding
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator

class MainActivity : AppCompatActivity() {
    var categoryAdapter: CategoryAdapter? = null
    var tabLayout: TabLayout? = null
    var viewPager2: ViewPager2? = null

    private lateinit var binding : ActivityMainBinding
    private val titles = arrayOf("Mountains", "Cities", "Islands", "Seas")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)

        viewPager2 = binding.viewPager
        tabLayout = binding.tabLayout
        categoryAdapter = CategoryAdapter(this)

        //set adapter viewpager
        viewPager2!!.adapter = categoryAdapter
        TabLayoutMediator(
            tabLayout!!,
            viewPager2!!
        ) { tab: TabLayout.Tab, position: Int -> tab.text = titles[position] }.attach()
    }
}